instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/utils/Utils.java

```diff
@@ -0,0 +1,108 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.utils;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.javaagent.instrumentation.undertow.UndertowHttpServerTracer;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.Headers;
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import org.hypertrace.agent.core.config.HypertraceConfig;
+import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
+import org.hypertrace.agent.core.instrumentation.SpanAndBuffer;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedBuffersFactory;
+import org.hypertrace.agent.core.instrumentation.buffer.BoundedByteArrayOutputStream;
+import org.hypertrace.agent.core.instrumentation.utils.ContentTypeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xnio.channels.StreamSourceChannel;
+
+public final class Utils {
+
+  private static final Logger log = LoggerFactory.getLogger(Utils.class);
+
+  /**
+   * Creates a {@link SpanAndBuffer} and stores it in the {@link
+   * io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext} with the {@link
+   * StreamSourceChannel} as the key so that later invocations of {@link StreamSourceChannel#read}
+   *
+   * @param httpServerExchange that {@link HttpServerExchange#getRequestChannel()} was invoked on
+   * @param returnedChannel {@link StreamSourceChannel} implementation created and held by {@link
+   *     HttpServerExchange} for the request body
+   * @param contextStore to hold the {@link SpanAndBuffer} associated with the {@link
+   *     StreamSourceChannel} returned by {@link HttpServerExchange#getRequestChannel()}
+   */
+  public static void createAndStoreBufferForSpan(
+      final HttpServerExchange httpServerExchange,
+      final StreamSourceChannel returnedChannel,
+      final ContextStore<StreamSourceChannel, SpanAndBuffer> contextStore) {
+    if (!HypertraceConfig.get().getDataCapture().getHttpBody().getRequest().getValue()
+        || !ContentTypeUtils.shouldCapture(
+            httpServerExchange.getRequestHeaders().getFirst(Headers.CONTENT_TYPE))) {
+      return;
+    }
+    final Span span =
+        Span.fromContext(UndertowHttpServerTracer.tracer().getServerContext(httpServerExchange));
+    final Charset charset = Charset.forName(httpServerExchange.getRequestCharset());
+    final BoundedByteArrayOutputStream boundedByteArrayOutputStream =
+        BoundedBuffersFactory.createStream(
+            (int) httpServerExchange.getRequestContentLength(), charset);
```

Comment:
what size is used if the content length is `0` e.g. for chunked transfer

Reply:
The `BoundedBuffersFactory` has some logic to help us out with that. In the case of stream being created in this way, we'll simply use the default `ByteArrayOutputStream` constructor which uses the default initial size of 32. Regardless, the ByteArrayOutputStream will grow to fit the size of its content as we write to it (until restricted by the `BoundedByteArrayOutputStream`). I'm sure we can improve on this, but I'm trying to keep this PR scoped to the bug with the customer as I don't have great data on how/when we can improve this right now